hibernate-core/src/main/java/org/hibernate/mapping/Column.java

```diff
@@ -435,9 +436,31 @@ public Size getColumnSize(Dialect dialect, Mapping mapping) {
 
 	Size calculateColumnSize(Dialect dialect, Mapping mapping) {
 		Type type = getValue().getType();
-		Long lengthToUse = getLength();
-		Integer precisionToUse = getPrecision();
-		Integer scaleToUse = getScale();
+		Long lengthToUse = null;
+		Integer precisionToUse = null;
+		Integer scaleToUse = null;
+		if ( sqlTypeName == null ) {
+			lengthToUse = getLength();
+			precisionToUse = getPrecision();
+			scaleToUse = getScale();
+		}
+		else {
+			// @Column(columnDefinition) present
+			int start = sqlTypeName.indexOf( '(' );
+			int end = sqlTypeName.indexOf( ')' );
+			if ( start > 0 && end > start + 1 && !sqlTypeName.toLowerCase( Locale.ROOT ).startsWith( "enum" ) ) {
+				String[] array = sqlTypeName.substring( start + 1, end ).split( ",", 2 );
+				try {
+					lengthToUse = Long.valueOf( array[0].trim() );
+					precisionToUse = Integer.valueOf( array[0].trim() );
+					if ( array.length == 2 ) {
+						scaleToUse = Integer.valueOf( array[1].trim() );
+					}
+				}
+				catch ( NumberFormatException ignore ) {
+				}
+			}
+		}
```

Comment:
I'm not keen on this at all, for the following reasons:

1. `@Column(columnDefinition)` doesn't specify a SQL type, but rather a freeform SQL fragment. Sure, it's quite _likely_ that the first thing you find in parentheses is likely to be a length, but there's really no guarantee of this at all.
2. We explicitly recommend against the use of `columnDefinition`, and in JPA 3.2 there is `options` as a much better alternative.
3. If a use _is_ using `columnDefinition` together with `SchemaUpdate`, then they can just explicitly specify `length` or `precision`.

Ultimately I think this is pretty fragile, and it's better to just steer people clear of `columnDefinition` (which we already do). I never liked it anyway, that was a TopLink thing.

Reply:
> Then `length` or `precision` should be specified twice?

If you expect `SchemaUpdate` to work? Yeah, sure, why not?

There's really no expectation that Hibernate knows how to parse an arbitrary native DDL fragment. We don't use `columnDefinition` for specifying SQL types, but for more complicated things.

<img width="998" alt="Screenshot 2024-04-10 at 11 13 39 AM" src="https://github.com/hibernate/hibernate-orm/assets/579974/7dadfc1d-cf98-4dcd-bcd7-4215acd9621c">

Reply:
OK, I'll leave it as is.